cuML/src/metrics/trustworthiness.cu

```diff
@@ -0,0 +1,162 @@
+        updateHost(h_pred_I, d_pred_I, n * n_neighbors, stream);
+
+        CUDA_CHECK(cudaFree(d_pred_I));
+        CUDA_CHECK(cudaFree(d_pred_D));
+        return h_pred_I;
+    }
+
+    namespace Metrics {
+
+        /**
+        * @brief Compute the trustworthiness score
+        * @input param X: Data in original dimension
+        * @input param X_embedded: Data in target dimension (embedding)
+        * @input param n: Number of samples
+        * @input param m: Number of features in high/original dimension
+        * @input param d: Number of features in low/embedded dimension
+        * @input param n_neighbors: Number of neighbors considered by trustworthiness score
+        * @return Trustworthiness score
+        */
+        template<typename math_t>
+        double trustworthiness_score(const cumlHandle& h, math_t* X,
+            math_t* X_embedded, int n, int m, int d, int n_neighbors)
+        {
+            const int TMP_SIZE = MAX_BATCH_SIZE * n;
+
+            cudaStream_t stream = h.getStream();
+            auto alloc = h.getHostAllocator();
+
+            constexpr auto distance_type = MLCommon::Distance::DistanceType::EucUnexpandedL2Sqrt;
+            size_t workspaceSize = 0; // EucUnexpandedL2Sqrt does not need any workspace
+            typedef cutlass::Shape<8, 128, 128> OutputTile_t;
+            bool bAllocWorkspace = false;
+
+            math_t* d_pdist_tmp;
+            allocate(d_pdist_tmp, TMP_SIZE);
+            int* d_ind_X_tmp;
+            allocate(d_ind_X_tmp, TMP_SIZE);
+            int* h_ind_X = (int*)alloc->allocate(n * n * sizeof(int), stream);
+
+            int toDo = n;
+            while (toDo > 0)
+            {
+                int batchSize = min(toDo, MAX_BATCH_SIZE);
+                // Takes at most MAX_BATCH_SIZE vectors at a time
+
+                MLCommon::Distance::distance<distance_type, math_t, math_t,
+                                                        math_t, OutputTile_t>
+                        (&X[(n - toDo) * m], X,
+                        d_pdist_tmp,
+                        batchSize, n, m,
+                        (void*)nullptr, workspaceSize,
+                        stream
+                );
+                CUDA_CHECK(cudaPeekAtLastError());
+
+                sortColumnsPerRow(d_pdist_tmp, d_ind_X_tmp,
+                                    batchSize, n,
+                                    bAllocWorkspace, NULL, workspaceSize,
+                                    stream);
+                CUDA_CHECK(cudaPeekAtLastError());
+
+                updateHost(&h_ind_X[(n - toDo) * n], d_ind_X_tmp,
+                                            batchSize * n, stream);
+
+                toDo -= batchSize;
+            }
+
+            long* ind_X_embedded = get_knn_indexes(h, X_embedded,
+                                    n, d, n_neighbors + 1);
+            
+            double t = 0.0;
+            for (size_t i = 0; i < n; i++)
```

Comment:
Since the construction of the trustworthiness score is associative and transitive with respect to each of the rows of the distance & knn matrices, we should be calculating the ranks for each row on device and using `atomicAdd` to sum them into `t`

Reply:
In fact, to maximize utilization, each thread can compute the rank for a single neighbor. That would make the parallelism a function of n*n_neighbors, rather than n.

Reply:
Thanks for the review, yes it would be a lot better to make use of Cuda here. I'll work on that.